### ISSUE TITLE
💬 RFC: ability to use branch information in fetch:template

### ISSUE THREAD

**Issue by punkle** [backstage/backstage#17647]
### 🔖 Need

The `fetch:template` github action provides the ability to retrieve a template from a url and template values into the template files. Oftentimes that url is hosted on some git repository in github. It would be super nice if it were possible to refer to the branch name or other metadata about the repository that created the template.

e.g. it might be possible to do the following:

```
${{ values.metadata.github_sha }}
```

### 🎉 Proposal

To get this conversation started, I thought it might be a good idea to extend the `ReadTreeResponse` to include arbitrary metadata and then pass that metadata through to the values in the template.

```
export type ReadTreeResponse = {
  ...
  + metadata?: Record<string, string>;
};
```

### 〽️ Alternatives

I am not sure of another way to do this at this time.

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by electrofelix** [comment]
This was triggered by a discussion where I was asking how would one determine the version so that it could be recorded into a resulting repo generated from a template to allow subsequent checking in the future to determine what had changed in the template since a repo creation.

While it's possible to add an entry and ensure it is updated each time the source to be rendered for a template is updated it would be much easier if it was possible to ensure it was going to be updated automatically in some way to track the source the resulting contents were generated from.

For github, since it already supports exposing repo contents as a zip archive, and when doing so github will use `.gitattributes` containing `export-subst` to allow for a field in a specific file to contain the git SHA1 the archive came from.

Done this before for vagrant-libvirt using https://github.com/vagrant-libvirt/vagrant-libvirt/blob/99ac32862147f05fd6b13ddfcb3da2fd4928dfc0/lib/vagrant-libvirt/version.rb#L10 combined with https://github.com/vagrant-libvirt/vagrant-libvirt/blob/99ac32862147f05fd6b13ddfcb3da2fd4928dfc0/.gitattributes so package managers pulling source archives directly would get the commit hash and version or branch name depending on what ref the archive was generated from by github.

Assuming there was actions to download such an archive and then unpack it to the workspace, to a path that can be referenced, the subsequent `fetch:template` could specify a subdirectory as needed. This would work then for any remote download from source control services that take care of this without needing to know source control was involved for the fetch template.

Doing it this way might be more useful in general?

Though I would admit from a raw consumption perspective, just being able to get the git sha of the repo referenced by the template url path would be more straight forward to consume.

**Comment by jhaals** [comment]
Thanks for the suggestion, I can see the value in an identifier to which version of the template that the new thing is scaffolded from is useful. I'm slightly worried that placing an databag on ReadTreeResponse is too open and might lead to people wanting to put more information in that which might not be available directly or relying heavily on some metadata fields always being there. Are there other alternatives to this one worth exploring?

**Comment by electrofelix** [comment]
I think we have the following possible list of options so far:
* Expose source identifier for templates loaded from a repository, could be a git sha, svn commit, etc, or timestamp for reading from a user projected example
* Provide an unzip action and combine with a download archive action to leverage existing source control provides ability to export source containing additional source markers (possible race condition where started using a template read from a branch where the template dir in the source archive pulled has been updated subsequently)
* Extend https://github.com/arhill05/backstage-plugin-scaffolder-git-actions#readme to make output retrievable (assuming it makes sense to execute against a template)

Unfortunately I don't know enough about backstage and scaffolder plugins to be sure I'm making sensible suggestions. My naive thought is that if the template is loaded from github, and the `fetch:template` is run against a specific directory, it should be possible to have something else execute against that same path just before to retrieve any source markers of interest and then those could be feed in to the next action to be used as parameters?

**Comment by freben** [comment]
Ping @benjdlambert

**Comment by benjdlambert** [comment]
Hey @electrofelix - I'm thinking that we should probably look at making some information about the template in the `fetch:template` action so that you can use those variables inside the template like the current sha that it downloaded from or whatnot. I'm not feeling that extending the `readTree` api with the information is the right change though, as `readTree` and `readUrl` etc might not always be reading from git, or an SCM, so it's probably not the right place to do it. 

Basically you would get something like `${{ meta.template.revision }}` or something that you could use inside your `fetch:template` templates? Does this make sense?

**Comment by electrofelix** [comment]
@benjdlambert yes, sorry, been busy and missed following up until I got the ping from the stalebot.

I think just as long as the information is enough to uniquely identify what code went into the template, that will be sufficient. I certainly understand that avoiding exposing too much from the underlying SCM makes sense.

What you are describing would definitely work for my usecase, hopefully others would find it useful as well.

**Comment by electrofelix** [comment]
Still relevant, hopefully triggers bot to remove stale label.

**Comment by benjdlambert** [comment]
Hey @SuganJoe :wave:

So I think that https://github.com/backstage/backstage/issues/17647#issuecomment-1547521888 is the most amount of information we have for an implementation here.

Happy to accept contributions for this of course and discuss further in a PR. Not sure that putting it on the `readTree` response is the right fix here as there isn't always a revision, but maybe if we can get some data from there and normalize it before passing it into the template could be a good idea.

Not sure what's going to be best, have a play around and see what you can come up with 👍
